Adds optional support for Docker/Compose

iMerica opened the pull request:
Hey guys,

Congrats on the dev.to launch! I've added some preliminary support for running the app locally using Docker/Compose. There are tons of advantages of running the app in Docker vs traditional local dev setups, for example:

- Being able to have one command (`docker-compose up -d`) to launch the entire stack locally. No dependencies required.
- Separating concerns in a more organized manner.
- More production-like application environment.
- Down the road, being able to deploy using containers. e.g `docker run practical-dev/devto:latest` ☁️ 

## What type of PR is this? (check all applicable)
- [X] Feature

## Description
Adds optional support for running dev.to using Docker.

## Related Tickets & Documents
Resolve #299 
## Added to documentation?
  - [x] readme

## What gif best describes this PR or how it makes you feel?
![swarmnado-357x627-30-1](https://user-images.githubusercontent.com/15793250/46823101-4f3e4880-cd5b-11e8-83b8-d71b04630464.gif)
Thanks,

Michael ([iMichael](https://dev.to/imichael) on dev.to)

rhymes commented on config/database.yml:
```diff
@@ -27,25 +27,25 @@ default: &default
   <<: *default
-  database: PracticalDeveloper_development
+  database: postgres
 
```
Hi @iMerica, I was wondering why you changed the default database name, is it just to make the build work with Docker?

Same question goes for the other params. You set a username which, though it's the default user installed with Postgres, might not be the newcomer first choice. Same reasoning goes with the password and the hostname. 

Thanks!

I feel like

iMerica replied:
These are the default credentials on the official [postgres](https://hub.docker.com/_/postgres/) container. I'll replace them with the originals using env vars.

bdod replied:
why you don't write some kind of entrypoint script, something like this ?
```
cat <<EOF > config/database.yml

default: &default
  adapter: postgresql
  encoding: unicode
  # For details on connection pooling, see rails configuration guide
  # http://guides.rubyonrails.org/configuring.html#database-pooling
  pool: 5
  connect_timeout: 2
  checkout_timeout: 2
  host:     <%= ENV['POSTGRES_HOST'] || 'postgres' %>
  username: <%= ENV['POSTGRES_USER'] || 'postgres' %>
  password: <%= ENV['POSTGRES_PASSWORD'] || 'mysecretpassword' %>
  variables:
      statement_timeout: <%= ENV['STATEMENT_TIMEOUT'] || 2500 %>
      
development:
  <<: *default
  database: <%= ENV['POSTGRES_DATABASE'] || 'PracticalDeveloper_development' %>

test:
  <<: *default
  database: <%= ENV['POSTGRES_DATABASE'] || 'PracticalDeveloper_test' %>

production:
  <<: *default
  database: <%= ENV['POSTGRES_DATABASE'] || 'PracticalDeveloper_production' %>

EOF

set_var() {
VAR_NAME=$1
VAR_VALUE=$2
sed -i "/:${VAR_NAME}/ s~default:.*~default: \"${VAR_VALUE}\"~" Envfile
}

VARS=$(cat Envfile | awk '/variable/{print $2}' | sed 's/[:|,]//g')
for v in $VARS ; do  [[ ${!v} ]] && set_var $v ${!v} ; done

```

iMerica replied:
This is not necessary. The PostgresSQL container supports overrides.

bdod replied:
And here postgres container ?, so you do not need to interfere with the settings or something else ... look at my MR #342

iMerica replied:
Hi @bdod,

> hardcoding settings it's a bad practice

Please see the [docs](https://hub.docker.com/_/postgres/) for the Postgres container. It is not "hardcoding" when by receiving postgres connection details from the environment.  This is a common 12-factor practice.

maestromac commented:
Hey @iMerica, Thanks for doing this! I dabbled into this  ~6 months ago; got it working but super slow and I end up left it collecting dust. I pushed up my work to [this commit](https://github.com/thepracticaldev/dev.to/commit/7f5b9a50ad1f3451771c6e412d3c7c8bca034f2f), maybe it would be of use to you.

rhymes commented:
@maestromac I guess I'm not the only one that's not using Docker because it's super slow on my Mac :-(

iMerica commented:
I'll push a working configuration to my branch today. For now, Docker will be an optional opt-in thing. e.g "Use the traditional local Rails setup or run the entire stack locally using Docker"

iMerica commented:
The only thing pending on this PR is that migrations fail due to the following error:

> Directly inheriting from ActiveRecord::Migration is not supported. Please specify the Rails release the migration was written for.

This probably doesn't impact existing developers because newer migrations use `ActiveRecord::Migration[5.1]`. Any ideas @maestromac ? 

Other notes:
- I've added a few more services to make the local dev setup more production-like:
  - delayed_job
  - redis
  - cable
- Looks like there is an undeclared dependency for delayed_job so I've added it (`daemons v1.2`)
- Rails can use `DATABASE_URL` which is a bit more flexible between environments so I've added that to the development config. Let me know if you guys have other preferences.

maestromac commented:
We've managed to avoid `Directly inheriting from ActiveRecord::Migration is not supported` by relying on `db:schema:load` (which is part of db:setup) instead of `db:migrate`. `db:migrate` is still used in subsequent migrations after initial `bin/setup`.

I have played around with your pr and I got it partially working by switching `db:migrate` to `db:schema:load`. I then hit this error.

```
...
-- add_foreign_key("messages", "chat_channels")
   -> 0.0086s
-- add_foreign_key("messages", "users")
   -> 0.0059s
-- add_foreign_key("push_notification_subscriptions", "users")
   -> 0.0092s
-- enable_extension("plpgsql")
rails aborted!
PG::ConnectionBad: could not connect to server: Connection refused
        Is the server running on host "localhost" (127.0.0.1) and accepting
        TCP/IP connections on port 5432?
could not connect to server: Cannot assign requested address
        Is the server running on host "localhost" (::1) and accepting
        TCP/IP connections on port 5432?
```

Also, would a dedicated `bin/webpack-dev-server` service be required?

iMerica commented:
@maestromac Thanks.  I should be able to fix the remaining issues then.

rhymes commented on config/environments/development.rb:
```diff
@@ -2,7 +2,7 @@
 Rails.application.configure do
   # Verifies that versions and hashed value of the package contents in the project's package.json
-  config.webpacker.check_yarn_integrity = true
+  config.webpacker.check_yarn_integrity = false
```
Thi should probably be restored once the PR is solid... For more information: https://github.com/rails/webpacker/blob/master/README.md#yarn-integrity

iMerica replied:
👍

iMerica replied:
On second thought this seems like an anti-pattern. 

The rails container/service should be concerned with running the rails dev server. The webpack service should be concerned with yarn integrity.

rhymes replied:
There is no "webpack service" though :D webpacker is a library for Rails, this checks if the yarn.lock and the package.json are aligned, the same way Bundler does for Gemfile and Gemfile.lock. I wouldn't worry about it too much. 

It's also the default of the library...

iMerica replied:
Blocked by this:

- https://github.com/rails/webpacker/issues/1568
- https://github.com/yarnpkg/yarn/issues/3167

Works fine when ` config.webpacker.check_yarn_integrity = false`

chuckha commented on config/database.yml:
```diff
@@ -24,9 +24,9 @@ default: &default
-      
 development:
   <<: *default
+  url: <%= ENV.fetch('DATABASE_URL') %>
```
This uses .fetch which [throws a KeyError](Your board will ship in 3-4 weeks.
) if no default is provided and the value doesn't exist. Will this have unexpected side effects for non-docker users?

iMerica replied:
Good catch 👍 . As part of testing this PR, we need to do a clean install for non-docker users.

chuckha commented on infra/dev/rails/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+
+RUN yarn
+
+FROM ruby:2.5.1 as ruby_build
```
It would be nice to build/install ruby based on the .ruby-version file so that it's not hard coded in two places. This feels more like a future task though.

iMerica replied:
Might be possible using build args 👍

chuckha commented on infra/dev/rails/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+
+ENV RAILS_ENV development
+
+RUN bundle check || bundle install --jobs 20 --retry 5
```
I know it's part of `bin/setup` but i think it's safe to not do `bundle check` first since no gems will ever exist beforehand

chuckha commented on docker-compose.yml:
```diff
@@ -0,0 +1,43 @@
+  migrate:
+    ports: []
+    <<: *rails_base
+    command: bin/rails db:migrate
```
Should there be a step for db:setup for the first run? Might be a documentation fix

chuckha commented:
What's left to do here? Need an extra set of hands or eyes? I would be happy to help in whatever way to push this along

iMerica commented:
Fixed all the backend plumbing, all that is left is minor frontend issues:

![image](https://user-images.githubusercontent.com/487897/44361841-61e38200-a47c-11e8-9da2-455444c65ad8.png)

I actually don't have the bandwidth to address these frontend issues. So if someone wants to take that (@chuckha ?) be my guest 👍. My only advice is to separate web-pack it as much as possible from Rails. In fact, I would suggest running the webpack dev server in developer's host operating system – not in Docker (I'm a docker fanatic and even I dislike running webpack dev servers in Docker).

iMerica commented on docker-compose.yml:
```diff
@@ -0,0 +1,48 @@
+    depends_on:
+      - 'redis'
+    command: puma -p 28080 cable/config.ru
+  webpacker:
```
I added this webpacker service, but maybe it would be better to run al this stuff in the host operating system?

chuckha commented:
nice, thanks for the update, I'll take a look this evening!

ryuheechul commented:
Can't wait to see this merged!

mhobotpplnet commented:
Agree, I want to spun this up in my K8s cluster.

benhalpern commented:
Our team's going to prioritize finishing this up this week.

maestromac commented:
Hey @iMerica, can you please sign our CLAassistant? I'm taking to take over this PR.

iMerica commented:
@maestromac Done 👍

maestromac commented:
I have went ahead and simplified the setting a bit. With the current implementation, it takes about 3min for initial homepage load on MacOS... but it works 😅. The most important thing is this PR should not affect native development. We'll merge this for now and are more than happy to accept any improvement PRs!

iMerica commented:
👍

rhymes commented:
Yay! Well done @maestromac :-) This will help all non-Ruby folks that want to contribute
